### PR TITLE
ACS AEM Commons Touch UI Multifield Extension

### PR DESCRIPTION
**Pull request by schoudry**
New clientlib "touchui-widgets" extending ootb "cq.authoring.dialog" clientlib providing the composite multifield capability

### CODE REVIEW

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+                    </column>
+                </items>
+            </content>
+        </jcr:root>
+     </code>
+*/
+
+
+(function () {
+    var DATA_ACS_AEM_NESTED = "data-acs-aem-nested",
```
Comment (justinedelson) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: Can we change this to data-acs-commons-nested ? the "aem" part is IMHO unnecessary (as we're obviously in AEM  :smile: )

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        CFFW = ".coral-Form-fieldwrapper",
+        _ = window._, CUI = window.CUI, Class = window.Class;
+
+    //reads multifield data from server, creates the nested composite multifields and fills them
+    function addDataInFields() {
+        $(document).on("dialog-ready", function() {
+            var mName = $("[" + DATA_ACS_AEM_NESTED + "]").data("name"),
```
Comment (justinedelson) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: not really a big deal, but there's three different naming conventions here - m<something>, $<something>, and just <something>. Is there a reason for this?

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js (hunk 1 above)
Comment (schoudry) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: sure, makes sense, and btw i am going to implement all review comments at one go, eod

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js (hunk 2 above)
Comment (schoudry) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: hmm, mName is a plain string, here its multifield name say "./pages" , a variable prefixed with $ is output of some jquery selector, you know to easily identify in the code, can apply jquery functions on the variable without doing $() again... i didn't get the "."

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+    $(document).ready(function () {
+        addDataInFields();
+        collectDataFromFields();
+    });
+
+    //extend otb multifield for adjusting event propagation when there are nested multifields
+    //for working around the nested multifield add and reorder
+    CUI.Multifield = new Class({
```
Comment (visiongeist) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: please do not attach that to the CUI namespace. You will not be happy about that in future ;)

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+            $('<input />').attr('type', 'hidden')
+                .attr('name', mName)
+                .attr('value', JSON.stringify(record))
+                .appendTo($form);
+        });
+    }
+
+    //collect data from widgets in multifield and POST them to CRX as JSON
+    function collectDataFromFields(){
+        $(document).on("click", ".cq-dialog-submit", submitAction);
```
Comment (visiongeist) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: You may also block the whole form submission in the capture phase. Might be safer than attaching to the click
document.querySelector('form.cq-dialog').addEventListener('submit', submitAction, true);

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js (hunk 3 above)
Comment (schoudry) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: sure ill change it

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js (hunk 4 above)
Comment (schoudry) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: sure

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js (hunk 1 above)
Comment (schoudry) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: Done

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js (hunk 3 above)
Comment (schoudry) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: Done

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js (hunk 4 above)
Comment (schoudry) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: Done

File: content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js (hunk 3 above)
Comment (justinedelson) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: Actually... If I understand this code (and I may not), it is working around a bug in CoralUI, in which case don't we *have* to call this CUI.Multifield (i.e. we need to replace CUI.Multifield with this extended version?)

The problem I'm forseeing is that somewhere else, someone needs to similarly monkey patch CUI.Multifield. If they do

{code}
new Class({
  extend: CUI.Multifield
...
}
{code}

They'll be extending the *original* CUI.Multifield, not this overridden class. So the changes made here would get lost.

Comment (schoudry) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: not sure about the bug but...

 yes, if they extend and call CUI.Widget.registry.register("multifield","other_cui_multifield_extension") the ACS_COMMONS_MULTIFIELD changes are lost

Comment (visiongeist) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: Ok I wasn't aware that you will only use your patched version. Alright then go ahead overriding the constructor.

Comment (schoudry) [inline content/src/main/content/jcr_root/apps/acs-commons/touchui-widgets/source/touchui-composite-multifield.js]: done

**Comment by justinedelson** [comment]
This looks good to me. Any other comments before merging?

**Comment by schoudry** [comment]
none from me

**Pull request closed without merging**